sdl/render.go

```diff
@@ -307,7 +307,8 @@ func (texture *Texture) Lock(rect *Rect) ([]byte, int, error) {
 
 	pitch := int32(_pitch)
 	if rect != nil {
-		length = int((pitch / w) * rect.W * rect.H)
+		bytesPerPixel := pitch / w
+		length = int(bytesPerPixel * (w*rect.H - rect.X - (w - rect.X - rect.W)))
```

Comment:
This calculation still looks pretty messy, sadly. I couldn't decide between more verbose and more compact.

A less minimal representation is 
```
length = int(bytesPerPixel*w*rect.H - bytesPerPixel*rect.X - (bytesPerPixel*(w - rect.X - rect.W))))
```